_refactor/combined.py

```diff
@@ -0,0 +1,589 @@
+            )
+            all_sites += sites
+        np.random.shuffle(all_sites)
+        return all_sites[:num_sites]
+
+    def _random_sites_on_triangle(self, surfsite_pos, k, added_z):
+        """
+        surfsite_pos: xyz coordinates of three points that forms a triangle in 3D space on surface
+        k: number of random sites to be generated
+        added_z: some extra cushion space in z direction for the random sites
+
+        """
+        A = np.c_[
+            surfsite_pos[:, 0], surfsite_pos[:, 1], np.ones(surfsite_pos.shape[0])
+        ]
+        coef, _, _, _ = scipy.linalg.lstsq(A, surfsite_pos[:, 2])
+        A, B, C = coef
+        rand_x = np.random.uniform(
+            low=np.min(surfsite_pos[:, 0]), high=np.max(surfsite_pos[:, 0]), size=(k,)
+        )
+        rand_y = np.random.uniform(
+            low=np.min(surfsite_pos[:, 1]), high=np.max(surfsite_pos[:, 1]), size=(k,)
+        )
+        rand_z = np.array(
+            [
+                A * sitex + B * sitey + C + added_z
+                for sitex, sitey in zip(rand_x, rand_y)
+            ]
+        )
+        sites = list(np.c_[rand_x, rand_y, rand_z])
+        return sites
+
+    def place_adsorbate_on_sites(self, adsorbate, surface, random_sites):
+        """
+        Place an adsorbates at a list of adsorption sites.
+        Args:
+            adsorbate : An `ase.Atoms` object of the adsorbate
+            surface : An `ase.Atoms` object of the surface
+            random_sites : A list of cartesian coordinates representing the random sites
+
+        Returns:
+            adslabs   A list of `ase.Atoms` object. Each object is an adsorbate placed
+                      on the surface
+        """
+        adslabs = []
+        for site in random_sites:
+            # Add adsorbate to a site
+            adsorbate_c = adsorbate.copy()
+            surface_c = surface.copy()
+            adsorbate_c.translate(site)
+            adslab = surface_c + adsorbate_c
+            tags = [2] * len(adsorbate)
+            final_tags = list(surface.get_tags()) + tags
+
+            # torsion rotate adsorbate if "rotate_adsorbate" is set to True
+            if self.rotate_adsorbate:
+                adsorbate_idx = [idx for idx, tag in enumerate(final_tags) if tag == 2]
+                struct = AseAtomsAdaptor.get_structure(adslab)
+                rotation = np.random.uniform(0, 2 * np.pi)
+                struct.rotate_sites(adsorbate_idx, rotation, (0, 0, 1), site)
+                adslab = AseAtomsAdaptor.get_atoms(struct)
+                adslab.set_tags(final_tags)
+
+            adslab.cell = surface.cell
+            adslab.pbc = [True, True, False]
+            adslabs.append((site, adslab))
+        return adslabs
+
+    def is_config_reasonable(self, adslab):
+        """
+        Function that check whether the adsorbate placement is reasonable.
+        Two criteria are: 1. The adsorbate should be placed on the slab:
```

Comment:
I dont see criteria 1 implemented?

Reply:
This seems to be a copy/paste of the heuristic `is_config_reasonable()` documentation. Is this criteria something we want to check for random placements, or was this intentionally taken out of the code?